File: ios/Classes/Registry/RegistryReader.swift
import Foundation

enum RegistryReaderError: Error {
    case invalidMagic
    case unsupportedVersion(UInt32)
    case decryptionFailed(String)
    case malformedData
}

enum RegistryReader {

    /// Reads and decrypts the fl_env binary registry.
    ///
    /// Binary format (all integers big-endian):
    ///   Bytes  Field
    ///   0-3    Magic: 0x464C454E ("FLEN")
    ///   4-7    Version: UInt32 = 1
    ///   8-11   Tier-1 entry count: UInt32
    ///   12-15  Tier-2 entry count: UInt32 (Phase 1: always 0)
    ///
    ///   Per entry:
    ///     key-length  UInt32
    ///     key         UTF-8 bytes
    ///     nonce       12 bytes
    ///     cipher-len  UInt32  (ciphertext + 16-byte GCM tag)
    ///     cipher+tag  bytes
    static func readAll(key: [UInt8], from data: Data) throws -> [String: String] {
        var offset = data.startIndex

        func readUInt32() throws -> UInt32 {
            guard offset + 4 <= data.endIndex else { throw RegistryReaderError.malformedData }
            let value = data[offset..<(offset + 4)].reduce(UInt32(0)) { ($0 << 8) | UInt32($1) }
            offset += 4
            return value
        }

        func readBytes(_ count: Int) throws -> [UInt8] {
            guard offset + count <= data.endIndex else { throw RegistryReaderError.malformedData }
            let bytes = Array(data[offset..<(offset + count)])
            offset += count
            return bytes
        }

        // Validate magic "FLEN"
        let magic = try readBytes(4)
        guard magic == [0x46, 0x4C, 0x45, 0x4E] else {
            throw RegistryReaderError.invalidMagic
        }

        let version = try readUInt32()
        guard version == 1 else {
            throw RegistryReaderError.unsupportedVersion(version)
        }

        let tier1Count = try readUInt32()
        _ = try readUInt32() // skip tier2Count (Phase 1: always 0)

        var result: [String: String] = [:]
        result.reserveCapacity(Int(tier1Count))

        for _ in 0..<tier1Count {
            let keyLen = Int(try readUInt32())
            let keyBytes = try readBytes(keyLen)
            guard let entryKey = String(bytes: keyBytes, encoding: .utf8) else {
                throw RegistryReaderError.malformedData
            }

            let nonce = try readBytes(12)
            let cipherLen = Int(try readUInt32())
            let cipherWithTag = try readBytes(cipherLen)

            let plaintext: [UInt8]
            do {
                plaintext = try AesGcmDecryptor.decrypt(key: key, nonce: nonce, cipherWithTag: cipherWithTag)
            } catch {
                throw RegistryReaderError.decryptionFailed(
                    "Decryption failed for key '\(entryKey)': \(error.localizedDescription)"
                )
            }

            guard let value = String(bytes: plaintext, encoding: .utf8) else {
                throw RegistryReaderError.malformedData
            }
            result[entryKey] = value
        }

        return result
    }
}

File: android/src/main/kotlin/com/pixmerc/fl_env/registry/RegistryReader.kt
package com.pixmerc.fl_env.registry

import android.content.Context
import com.pixmerc.fl_env.crypto.AesGcmDecryptor
import java.nio.ByteBuffer
import java.nio.ByteOrder

/**
 * Reads and decrypts the fl_env binary registry from `res/raw/fl_env_registry`.
 *
 * Binary format (all integers big-endian):
 *   Bytes  Field
 *   0-3    Magic: 0x464C454E ("FLEN")
 *   4-7    Version: UInt32 = 1
 *   8-11   Tier-1 entry count: UInt32
 *   12-15  Tier-2 entry count: UInt32 (Phase 1: always 0)
 *
 *   Per entry:
 *     key-length  UInt32
 *     key         UTF-8 bytes
 *     nonce       12 bytes
 *     cipher-len  UInt32  (ciphertext + 16-byte GCM tag)
 *     cipher+tag  bytes
 */
internal class RegistryReader(private val context: Context) {

    @Volatile
    private var cached: Map<String, String>? = null

    fun readAll(key: ByteArray): Map<String, String> {
        cached?.let { return it }

        val resId = context.resources.getIdentifier("fl_env_registry", "raw", context.packageName)
        check(resId != 0) { "fl_env_registry.bin not found in res/raw — run 'fl_env build' first." }

        val bytes = context.resources.openRawResource(resId).use { it.readBytes() }
        val result = parse(key, bytes)
        cached = result
        return result
    }

    /** Parse from raw bytes — used in unit tests without a real [Context]. */
    internal fun readFrom(key: ByteArray, bytes: ByteArray): Map<String, String> = parse(key, bytes)

    private fun parse(key: ByteArray, bytes: ByteArray): Map<String, String> {
        val buf = ByteBuffer.wrap(bytes).order(ByteOrder.BIG_ENDIAN)

        // Validate magic "FLEN"
        val magic = ByteArray(4).also { buf.get(it) }
        check(magic.contentEquals(byteArrayOf(0x46, 0x4C, 0x45, 0x4E))) {
            "fl_env registry has invalid magic header — file may be corrupt."
        }

        val version = buf.int
        check(version == 1) { "fl_env registry version $version is not supported." }

        val tier1Count = buf.int
        buf.int // skip tier2Count (Phase 1: always 0)

        val result = HashMap<String, String>(tier1Count)
        repeat(tier1Count) {
            val keyLen = buf.int
            val keyBytes = ByteArray(keyLen).also { buf.get(it) }
            val entryKey = String(keyBytes, Charsets.UTF_8)

            val nonce = ByteArray(12).also { buf.get(it) }

            val cipherLen = buf.int
            val cipherWithTag = ByteArray(cipherLen).also { buf.get(it) }

            val plaintext = AesGcmDecryptor.decrypt(key, nonce, cipherWithTag)
                ?: error("fl_env: decryption failed for key '$entryKey' — key mismatch or tampered data.")

            result[entryKey] = String(plaintext, Charsets.UTF_8)
        }

        return result
    }
}
